orderbook/usecase/orderbook_usecase_test.go

```diff
@@ -24,6 +31,260 @@ func TestOrderbookUsecaseTestSuite(t *testing.T) {
 	suite.Run(t, new(OrderbookUsecaseTestSuite))
 }
 
+func (s *OrderbookUsecaseTestSuite) TestProcessPool() {
+	testCases := []struct {
+		name          string
+		pool          sqsdomain.PoolI
+		setupMocks    func(usecase *orderbookusecase.OrderbookUseCaseImpl, client *mocks.OrderbookGRPCClientMock, repository *mocks.OrderbookRepositoryMock)
+		expectedError string
+	}{
+		{
+			name:          "pool is nil",
+			pool:          nil,
+			expectedError: "pool is nil when processing order book",
+		},
+		{
+			name: "cosmWasmPoolModel is nil",
+			pool: &mocks.MockRoutablePool{
+				CosmWasmPoolModel: nil,
+			},
+			expectedError: "cw pool model is nil when processing order book",
+		},
+		{
+			name: "pool is not an orderbook pool",
+			pool: &mocks.MockRoutablePool{
+				ID:                1,
+				CosmWasmPoolModel: &cosmwasmpool.CosmWasmPoolModel{},
+			},
+			expectedError: "pool is not an orderbook pool 1",
+		},
+		{
+			name: "orderbook pool has no ticks, nothing to process",
+			pool: &mocks.MockRoutablePool{
+				CosmWasmPoolModel: &cosmwasmpool.CosmWasmPoolModel{
+					ContractInfo: cosmwasmpool.ContractInfo{
+						Contract: cosmwasmpool.ORDERBOOK_CONTRACT_NAME,
+						Version:  cosmwasmpool.ORDERBOOK_MIN_CONTRACT_VERSION,
+					},
+					Data: cosmwasmpool.CosmWasmPoolData{
+						Orderbook: &cosmwasmpool.OrderbookData{
+							Ticks: []cosmwasmpool.OrderbookTick{},
+						},
+					},
+				},
+			},
+			expectedError: "",
+		},
+		{
+			name: "failed to cast pool model to CosmWasmPool",
+			pool: &mocks.MockRoutablePool{
+				CosmWasmPoolModel: &cosmwasmpool.CosmWasmPoolModel{
+					ContractInfo: cosmwasmpool.ContractInfo{
+						Contract: cosmwasmpool.ORDERBOOK_CONTRACT_NAME,
+						Version:  cosmwasmpool.ORDERBOOK_MIN_CONTRACT_VERSION,
+					},
+					Data: cosmwasmpool.CosmWasmPoolData{
+						Orderbook: &cosmwasmpool.OrderbookData{
+							Ticks: []cosmwasmpool.OrderbookTick{
+								{TickId: 1},
+							},
+						},
+					},
+				},
+				ChainPoolModel: &mocks.ChainPoolMock{
+					ID:   1,
+					Type: poolmanagertypes.Balancer,
+				},
+			},
+			expectedError: "failed to cast pool model to CosmWasmPool",
+		},
+		{
+			name: "failed to fetch ticks for pool",
+			pool: &mocks.MockRoutablePool{
+				CosmWasmPoolModel: &cosmwasmpool.CosmWasmPoolModel{
+					ContractInfo: cosmwasmpool.ContractInfo{
+						Contract: cosmwasmpool.ORDERBOOK_CONTRACT_NAME,
+						Version:  cosmwasmpool.ORDERBOOK_MIN_CONTRACT_VERSION,
+					},
+					Data: cosmwasmpool.CosmWasmPoolData{
+						Orderbook: &cosmwasmpool.OrderbookData{
+							Ticks: []cosmwasmpool.OrderbookTick{
+								{TickId: 1},
+							},
+						},
+					},
+				},
+				ChainPoolModel: &cwpoolmodel.CosmWasmPool{},
+			},
```

Comment:
Observing same structs to keep repeating. Consider defining a default mock routable pool and using a decorator pattern to modify fields that are necessary for the tests:
```go
func withPoolID(mock *mocks.RoutablePool, poolID uin64) *mocks.RoutablePool {
   mock.PoolID = poolID
   return mock
}
```

Reply:
I have moved forward with your suggested approach an implemented wrappers, it looks much cleaner now. :+1: